src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                    i,
+                    ref receiverStored,
+                    ref rewrittenReceiver,
+                    ref temps,
+                    ref positionsAssignedToTemp);
+
+                visitedArgumentsBuilder.Add(VisitExpression(arguments[i]));
+
+                foreach (var placeholder in argumentPlaceholders)
+                {
+                    RemovePlaceholderReplacement(placeholder);
+                }
+            }
+
+            Debug.Assert(temps?.Count is null or > 0);
+            return visitedArgumentsBuilder.ToImmutableAndFree();
+
+            void ensureTempTrackingSetup([NotNull] ref ArrayBuilder<LocalSymbol>? temps, ref BitVector positionsAssignedToTemp)
+            {
+                if (temps == null)
+                {
+                    temps = ArrayBuilder<LocalSymbol>.GetInstance();
+                    positionsAssignedToTemp = BitVector.Create(arguments.Length);
+                }
+            }
+
+            ImmutableArray<BoundInterpolatedStringArgumentPlaceholder> setupInterpolationDataIfNeeded(
+                int argumentIndex,
+                ref bool receiverStored,
+                ref BoundExpression? visitedReceiver,
+                ref ArrayBuilder<LocalSymbol>? temps,
+                ref BitVector positionsAssignedToTemp)
+            {
+                var argument = arguments[argumentIndex];
+
+                if (argument is BoundConversion { ConversionKind: ConversionKind.InterpolatedStringHandler, Operand: BoundInterpolatedString operand })
+                {
+                    // Handler conversions are not supported in expression lambdas.
+                    Debug.Assert(!_inExpressionLambda);
+                    var interpolationData = operand.InterpolationData.GetValueOrDefault();
+                    var creation = (BoundObjectCreationExpression)interpolationData.Construction;
+
+                    if (creation.Arguments.Length > (interpolationData.HasTrailingHandlerValidityParameter ? 3 : 2))
+                    {
+                        Debug.Assert(!((BoundConversion)argument).ExplicitCastInCode);
+
+                        // We have an interpolated string handler conversion that needs context from the surrounding arguments. We need to store
+                        // all arguments up to and including the last argument needed by this interpolated string conversion into temps, in order
+                        // to ensure we're keeping lexical ordering of side effects.
+                        ensureTempTrackingSetup(ref temps, ref positionsAssignedToTemp);
+                        Debug.Assert(!positionsAssignedToTemp.IsNull);
+
+                        foreach (var placeholder in interpolationData.ArgumentPlaceholders)
+                        {
+                            // Replace each needed placeholder with a sequence of store and evaluate the temp.
+                            var argIndex = placeholder.ArgumentIndex;
```

Comment:
Consider asserting that `argIndex < argumentIndex` (enforcing that `visitedArgumentsBuilder[argIndex]` usages below only reference already visited arguments)